### DISCUSSION TITLE
Tag Values to Telegram

### DISCUSSION THREAD

**Discussion by Welsyntoffie** [dmroeder/pylogix#183] [Ideas]
I have been reading about telegram bots in python and how to use them to send different replies with different commands.
It would be great of I could command /vibrations and then get a reply about all the motor bearing vibrations on the 3 PLC's that controls motors and pumps.
Any way you could setup an example of of such a scenario, get 40 tags from 3 PLC's and send it via telegram?

**Comment by TheFern2** [comment]
My suggestion to you is to read telegram documentation and create a bot and write commands, etc. Then play with pylogix separately. Once you know both api's, put that knowledge together and create your telegram/pylogix bot. Asking for someone to create a bot for you is a bit lazy imo especially when both pylogix/telegram have tons of examples.

**Reply by Welsyntoffie** [reply]
I have a couple of Arduino's controlled by my raspberry running domoticz and dashticz with dtgbot (telegram). I have a vpn on the raspberry and my phone to control my smart home from anywhere in the world. I DID THAT, by myself.

I am NO expert on coding but I manage to get by. I never asked you to create the bot. I merely asked for an example of how one would start reading tags once the bot received a command from the server and apply it in a text to send via telegram. Currently I am using Bash scripts and LUA.

To start playing around with this, I need to buy another PI and wait for my compactlogix L35 to arrive.

Most of the examples I have seen connects to 1 PLC and returns the values. I want to connect to 3 within the same session.

Plus the reason why I asked for an example is because the devs here are, well software developers, and coding should come as easy as insulting people right?

YOU DON'T KNOW ME, NEVER CALL ME LAZY!!

**Reply by TheFern2** [reply]
I didn't say you were lazy, just that it looks lazy because of the abundant information online for both api's. If you need three connections nothing changes, you need three PLC objects with their respective PLC IPs, you can't use one PLC object for multiple PLC's if that's what you were asking indirectly. My suggestion still stands create your bot, and work with one PLC at time, then build upon that. If you have problems with that specific scenario post your code, and or open a discussion/issue. Just because we're software developers doesn't mean we're here to write specific examples for every single application out there.

**Comment by dmroeder** [comment]
Alright, everyone calm down.  Let's just chalk this e-communication doesn't convey emotion very well.  Starting over...

A very simple example of reading from 3 PLC's (assuming the tag names are the same in each):

```python
import pylogix

tags = ["Tag1", "Tag2", "Tag3"]

plc1 = pylogix.PLC("192.168.1.10")
plc2 = pylogix.PLC("192.168.1.20")
plc3 = pylogix.PLC("192.168.1.30")

vals1 = plc1.Read(tags)
vals2 = plc2.Read(tags)
vals3 = plc3.Read(tags)

print(vals1)
print(vals2)
print(vals3)

plc1.Close()
plc2.Close()
plc3.Close()
```

**Reply by TheFern2** [reply]
> Any way you could setup an example of of such a scenario, get 40 tags from 3 PLC's and send it via telegram?

All calmed here. I gathered from the OP that he wanted a full telegram example which is why I pointed him to telegram docs.

**Reply by dmroeder** [reply]
I understand, that is how it sounded to me too.

@Welsyntoffie, let me know if that example helps you out.

**Comment by Welsyntoffie** [comment]
> Any way you could setup an example.

I assume it was the word SETUP that caused the confusion. Apologies for the incorrect phrasing of my question. 

> @Welsyntoffie, let me know if that example helps you out.

Yes thank you. Makes a ton of sense.

Instead of print(vals1) I would...
curl --data 'chat_id='$SendMsgTo --data-urlencode 'text='"$vals1" 'https://api.telegram.org/bot'$TelegramBotToken'/sendMessage'

Now that I have 1 foot in the door, I will wait for my devices to arrive then tackle it head on.

**Reply by dmroeder** [reply]
Keep in mind, vals1 will be a list of the Response class (.TagName, .Value, .Status). so you will be passing all of that to telegram.  You may/may not want that exactly, but  it will get you started.  I guess what I'm getting at is the response data might need to be formatted prior to sending it via telegram.  I guess when you actually start getting into the weeds, post here how it goes or you can email me directly if that is easier.

**Comment by Welsyntoffie** [comment]
I installed telebot (on the raspberry I did not want to use) with a working example of diceyclock.
I use the diceyclock example to call pylogix example 5 script instead of sending a random integer.

This is the output in telegram... My custom description with the tag value. Looks perfect
```
-1st Line: 3529.29
-2nd line: 11
```

This is the modified example.
So from what I gather a number can not be joined to a string. The number has to be converted to a string first then joined to other strings.
Is this fine as it is?

```
import telepot
from telepot.loop import MessageLoop
import sys
sys.path.append('..')

bot = telepot.Bot('**********:***********************************')

from pylogix import PLC

tag = [0,1,2]

tag_list = ['devide', 'TestDateTime.Second']
z = 0

with PLC() as comm:
    comm.IPAddress = '172.22.2.61'
    ret = comm.Read(tag_list)
    for r in ret:
        tag[z] = r.Value
        z = z + 1

stringval = str(round(tag[0],2))
stringval1 = str(tag[1])

ResultString="-1st Line: " + stringval
ResultString+='\n'
ResultString+="-2nd Line: " + stringval1

bot.sendMessage(chat_id, ResultString)
```

**Reply by dmroeder** [reply]
You can use string formatting to simplify, if that is what you are asking.  Format will convert to string for you.  Plus, you can do the rounding as well.  :.2f is rounding to 2 decimal places.

```python
val1 = 10.45678
val2 = 20

ResultString = "-1st Line: {:.2f}\n-2nd Line: {}".format(val1, val2)
```

**Comment by Welsyntoffie** [comment]
Just an update. 
I have a bunch of tags from multiple PLC's looking for state changes and sending to different chat id's depending on what tag changed.
Had a hiccup with telegram disconnected after a long idle because pylogix is looping trough the PLC's, but that got sorted out.
Implemented a kind of heartbeat signal aswell. A timer on one PLC will expire after 2 minutes and indicate on the scada system that the loop script is no more running, unless pylogix resets the timer before the 2 minutes.  

@dmroeder , thank you for the valuable advice. 
@TheFern2 ,thanks to you to, if all else fails read the instruction manual.

**Reply by TheFern2** [reply]
Nice work!

**Comment by Welsyntoffie** [comment]
O no. Removing the ethernet from a plc stops the script, but that caused by telepot because of the None value.

Is the word None the same as NO Comms? I assume now I should be testing for comms first before proceeding to the other parts of the script?

**Reply by dmroeder** [reply]
Instead of checking .Value != None, check that .Status == "Success"

Edit: fixed typo from typing on phone

**Reply by TheFern2** [reply]
You can also use the ping3 library, but yeah Status might be easier to do.

https://github.com/kyan001/ping3

**Reply by Welsyntoffie** [reply]
This works nice for cleaning up the code...
`ResultString = "-1st Line: {:.2f}\n-2nd Line: {}".format(val1, val2)`

I also tried to use the .Status to block certain code from running, with this I have to create if statements in other places.
Instead I opted for...
``` p = 0
    for r in ret:
          tag3[p] = r.Value
          if tag3[p] == None:
              tag3[p] = 999
            p = p + 1
```
Now I don't need to change code in different other places with if statements. A value off 999 is impossible since my pumps and motor vibrations trips at 8 and valve position can only be max 100%.
